set is_allocated in all horiz_interp_type_new routines

rem1776 opened the pull request:
**Description**
Sets the `is_allocated` field as well as the `method_type` within horiz_interp_types when called from any of the method specific horiz_interp_new routines. This should prevent reallocation errors as described in the issue below.

Also adds some calls to the unit test to test this functionality and renames one of the CPP macros to make more sense.

Fixes #1533

**How Has This Been Tested?**
make distcheck with oneapi 2024.01 on amd box

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

thomas-robinson commented on test_fms/horiz_interp/test_horiz_interp.F90:
```diff
@@ -979,6 +984,10 @@ subroutine test_assignment()
         real(HI_TEST_KIND_) :: D2R = real(PI,HI_TEST_KIND_)/180._lkind
         real(HI_TEST_KIND_) :: R2D = 180._lkind/real(PI,HI_TEST_KIND_)
         real(HI_TEST_KIND_), parameter :: SMALL = 1.0e-10_lkind
+        !! for bicubic set up
+        real(HI_TEST_KIND_), allocatable :: lon_src_1d(:), lat_src_1d(:)
+        real(HI_TEST_KIND_), allocatable :: lon_dst_1d(:), lat_dst_1d(:)
```
These should have doxygen documentation instead of just a comment above

rem1776 commented:
@thomas-robinson @bensonr This should be ready now, I updated the SPHERICA to SPHERICAL. I noticed that the interpolation method values were made public in the libFMS.F90 file so I also removed them from there as well since they should only be internal.
